src/core/lib/event_engine/cf_engine/dns_service_resolver.cc

```diff
@@ -0,0 +1,205 @@
+    on_resolve(std::move(result));
+    return;
+  }
+
+  DNSServiceRef sdRef;
+  auto host_string = std::string{host};
+  auto error = DNSServiceGetAddrInfo(
+      &sdRef, kDNSServiceFlagsTimeout | kDNSServiceFlagsReturnIntermediates, 0,
+      kDNSServiceProtocol_IPv4 | kDNSServiceProtocol_IPv6, host_string.c_str(),
+      &DNSServiceResolverImpl::ResolveCallback, this /* do not Ref */);
+
+  if (error != kDNSServiceErr_NoError) {
+    on_resolve(absl::UnknownError(
+        absl::StrFormat("DNSServiceGetAddrInfo failed with error:%d", error)));
+    return;
+  }
+
+  grpc_core::ReleasableMutexLock lock(&request_mu_);
+
+  error = DNSServiceSetDispatchQueue(sdRef, queue_);
+  if (error != kDNSServiceErr_NoError) {
+    on_resolve(absl::UnknownError(absl::StrFormat(
+        "DNSServiceSetDispatchQueue failed with error:%d", error)));
+    return;
+  }
+
+  requests_.try_emplace(
+      sdRef, DNSServiceRequest{
+                 std::move(on_resolve), static_cast<uint16_t>(port), {}});
+}
+
+/* static */
+void DNSServiceResolverImpl::ResolveCallback(
+    DNSServiceRef sdRef, DNSServiceFlags flags, uint32_t interfaceIndex,
+    DNSServiceErrorType errorCode, const char* hostname,
+    const struct sockaddr* address, uint32_t ttl, void* context) {
+  GRPC_EVENT_ENGINE_DNS_TRACE(
+      "DNSServiceResolverImpl::ResolveCallback: sdRef: %p, flags: %x, "
+      "interface: %d, errorCode: %d, hostname: %s, addressFamily: %d, ttl: "
+      "%d, "
+      "this: %p",
+      sdRef, flags, interfaceIndex, errorCode, hostname, address->sa_family,
+      ttl, context);
+
+  // no need to increase refcount here, since ResolveCallback and Shutdown is
+  // called from the serial queue and it is guarenteed that it won't be called
+  // after the sdRef is deallocated
+  auto that = static_cast<DNSServiceResolverImpl*>(context);
+
+  grpc_core::ReleasableMutexLock lock(&that->request_mu_);
+  auto request_it = that->requests_.find(sdRef);
+  GPR_ASSERT(request_it != that->requests_.end());
+  auto& request = request_it->second;
+
+  if (errorCode != kDNSServiceErr_NoError &&
+      errorCode != kDNSServiceErr_NoSuchRecord) {
+    request.on_resolve_(absl::UnknownError(absl::StrFormat(
+        "address lookup failed for %s: errorCode: %d", hostname, errorCode)));
+    that->requests_.erase(request_it);
+    DNSServiceRefDeallocate(sdRef);
+    return;
+  }
+
+  // set received ipv4 or ipv6 response, even for kDNSServiceErr_NoSuchRecord
+  if (address->sa_family == AF_INET) {
+    request.has_ipv4_response_ = true;
+  } else if (address->sa_family == AF_INET6) {
+    request.has_ipv6_response_ = true;
+  }
+
+  // collect results if there is no error (not kDNSServiceErr_NoSuchRecord)
+  if (errorCode == kDNSServiceErr_NoError) {
+    request.result_.emplace_back(address, address->sa_len);
+    auto& resolved_address = request.result_.back();
+    if (address->sa_family == AF_INET) {
+      ((struct sockaddr_in*)resolved_address.address())->sin_port =
```

Comment:
Please use C++-style casts, the C-style cases have mostly been removed from the codebase.

Reply:
Is it OK to const cast the return of `ResolvedAddress::address()`, otherwise it's going to be considerably more lines to create the addresses separately.

Reply:
The cast you're talking about is pretty common already.

https://github.com/grpc/grpc/blob/498fc99479a01177aa2ad257f1390521495b0788/src/core/lib/event_engine/tcp_socket_utils.cc#L223-L224

There is a clang tidy check enabled to try to prevent C-style casts from ending up in the codebase. I'm guessing tidy does not run on iOS code, but we should try to avoid them regardless. https://clang.llvm.org/extra/clang-tidy/checks/google/readability-casting.html

See ec0b88a5d73445ce5e6ff63c4691956bfe37e57f  and 2ff84ed8c2cb272b26766039d5bb07b009df2617 and d2bff16c25d7c207ef0afb6844fc50f567a2b36d and e0d8c498a60be0b5b5891e280bd441ec4a9d0785 for some of the work done here.